docs/txp.md

```diff
@@ -0,0 +1,459 @@
+
+## Overview
+
+Transaction processing consists of two parts: global and local. Global
+transaction processing is a part of block processing which checks
+whether transactional payload of blocks is valid and updates
+corresponding state when blocks are applied/rolled back. Local
+transaction processing checks standalone transactions and updates
+local mempool. These two parts have almost same logic with a few
+differences. We first describe global transaction processing and then
+describe how local transaction processing differs from it.
+
+Transaction processing is abbreviated to `txp` in code and some other
+places.
+
+## Global transaction processing
+
+Global transaction processing can be described by presenting an
+algorithm to solve the following problem:
+
+> Given a sequence of blocks `B₀, B₁, B₂, …` (where `B₀` is the first genesis block) check whether
+transactions from these blocks are valid.
+
+The algorithm is similar to the one from Bitcoin (for example, UTXO is used to prevent
+double-spending), but is more complicated due to various reasons, including:
+
+* Extendable (via update mechanism) data structures
+* Maintaining stakes in addition to balances
+
+
+We describe global txp as stateful algorithm:
+* Initial state `S₀` is derived from blockchain genesis data (see [mainnet genesis JSON](https://raw.githubusercontent.com/input-output-hk/cardano-sl/e7cfb1724024e0db2f25ddd2eb8f8f17c0bc497f/node/mainnet-genesis.json))
+* `S₁, S₂, …` are maintained as sequential application of blocks  `B₀, B₁, B₂, … ` to state `S₀`
+* State transition function. Given GState `S` and a main block `B` return either an error describing why `B` is invalid (w.r.t. tx payload) or new GState `S'`
+  ```
+  verifyAndApplyGState :: GState -> MainBlock -> Either TxPayloadVerificationError GState
+  -- ^ Note, that function definition and types are different in code and are put here for reader's convenience
+  ```
+  * Note, that state transition function is defined only for main blocks. This is done because genesis blocks (as defined in Ouroboros) don't have tx payload and thus are not considered by global txp.
+  * Recall that tx payload contains transactions (stored in Merkle tree) and their witnesses.
+
+Note, that in theory transaction verification can be stateless (except genesis
+state `S₀`) and be described without mentioning any additional state
+(apart from blocks themselves). But in practice it would be very
+inefficient and inconvenient.
+
+For sake of simplicity, we describe state transition function in two parts:
+* Verification: given GState `S` and a main block `B`, check whether `B` is valid (and can be applied to `S`)
+* Modification: given GState `S` and a main block `B` (successfully verified against `S`), produce `S'`
+
+### GState
+
+In this section we describe parts of GState relevant to transaction processing (GState is an actual structure in code used by few more components).
```

Comment:
What do you mean by `actual structure in code`? There is no `GState` type in code (or anything equivalent or similar).